File: cpp/bumped.cpp
// #include <bits/stdc++.h>
// using namespace std;

#include <bits/stdc++.h>
using namespace std;

class WeightedGraph {
   private:
    int edges;
    unordered_map<int, set<int>> adjacencyList;
    set<int> vertices;

   public:
    void addVertex(int v) {
        bool changed = vertices.find(v) == vertices.end();
        vertices.emplace(v);
        if (changed) {
            adjacencyList[v] = {};
        }
    }
};
int main() {}
// // Structure to represent an edge with weight
// struct Edge {
//     int destination;
//     int weight;
//     Edge(int dest, int w) : destination(dest), weight(w) {}
//     bool operator>(Edge& e) { return weight > e.weight; }
// };
// struct Flight {
//     int to;
//     int from;
//     Flight(int from, int to) : from(from), to(to){};
// };

// class WeightedGraph {
//    private:
//     int numVertices;
//     int numNodes;
//     vector<vector<Edge>> graph;
//     unordered_map<int, int> dist;

//    public:
//     WeightedGraph(int numV, int numNodes) : numVertices(numV), graph(numV), dist(numNodes) {}

//     void addEdge(int source, int destination, int weight) {
//         graph[source].push_back(Edge(destination, weight));
//         graph[destination].push_back(Edge(source, weight));  // Add the reverse edge
//         dist[source] = INT_MAX;
//         dist[destination] = INT_MAX;
//     }

//     void updDist(int node, int w) { dist[node] = w; }

//     vector<Edge> getNeighbours(int node) { return graph[node]; }
// };

// vector<Edge> dijkstra(WeightedGraph& g, Edge source, Edge goal) {
//     set<Edge> visited = {source};
//     priority_queue<Edge, vector<Edge>> q;
//     unordered_map<Edge, Edge> prev;
//     q.push(source);
//     g.updDist(source.destination, 0);
// }

// int main() {
//     int n, m, f, s, t;
//     cin >> n >> m >> f >> s >> t;
//     WeightedGraph graph(m, n);
//     vector<Flight> flights;

//     for (int i = 0; i < m; i++) {
//         int x, y, c;
//         cin >> x >> y >> c;

//         graph.addEdge(x, y, c);
//     }

//     for (int i = 0; i < f; i++) {
//         int x, y;
//         cin >> x >> y;
//         flights.push_back(Flight(x, y));
//     }

//     return 0;
// }

File: python/bumped.py
import math
import heapq


class WeightedGraph:
    def __init__(self, vertices, edges):
        self.numVertices = vertices
        self.numEdges = edges
        self.vertices = []
        self.adj = {}
        self.edges = {}

    def addNode(self, node: int):
        if node not in self.adj:
            self.adj[node] = set()
            self.vertices.append(node)

    def addEdge(self, a: int, b: int, weight: int):
        self.adj[a].add(b)
        self.adj[b].add(a)
        if self.edges.get((a, b), -1) == -1 and self.edges.get((b, a), -1) == -1:
            self.edges[(a, b)] = weight
            self.edges[(b, a)] = weight

    def getWeight(self, a: int, b: int):
        return self.edges.get((a, b), 0)

    def getNeighbours(self, node: int):
        return self.adj[node]

    def getVertices(self):
        return self.vertices


def dijkstra(graph, root, goal):
    dist = {}
    dist[root] = 0
    prev = {}
    seen = set()
    queue = []

    for n in graph.getVertices():
        if n != root:
            dist[n] = math.inf
            prev[n] = None
        heapq.heappush(queue, (dist[n], n))

    while queue:
        _, curr = heapq.heappop(queue)

        if curr in seen:
            continue

        seen.add(curr)

        for n in graph.getNeighbours(curr):
            newDist = dist[curr] + graph.getWeight(curr, n)

            if newDist < dist[n]:
                dist[n] = newDist

            heapq.heappush(queue, (newDist, n))
            prev[n] = curr
            # if n == goal:
            #     return dist, prev
    return dist, prev


def main():
    n, m, f, s, t = map(int, input().split())
    graph = WeightedGraph(n, m)
    flights = []

    for _ in range(m):
        i, j, c = map(int, input().split())
        graph.addNode(i)
        graph.addNode(j)
        graph.addEdge(i, j, c)

    for _ in range(f):
        i, j = map(int, input().split())
        flights.append((i, j))

    dist, prev = dijkstra(graph, s, t)

    # do dijkstra for whole graph
    # do dijkstra on all pairs of flights
    # find most expensive flights
    #
    # maxFlight = 0
    # flightDest = 0
    # flightFrom = 0
    # for x, y in flights:
    #     dist1, prev1 = dijkstra(graph, x, y)
    #     if maxFlight < dist1[y]:
    #         maxFlight = dist1[y]
    #         flightDest = y
    #         flightFrom = x

    # dist2, prev2 = dijkstra(graph, flightDest, t)

    # print(dist[flightFrom])
    # print(maxFlight)
    # print(dist2[t])
    # print(dist[flightFrom] + dist2[t] - maxFlight)


if __name__ == "__main__":
    main()
